### DISCUSSION TITLE
Developing a WordPress Plugin

### DISCUSSION THREAD

**Discussion by hirasso** [humbug/php-scoper#1084] [Q&A] [answered]
I am making a WordPress plugin to generate [ThumbHash](https://github.com/evanw/thumbhash) placeholders for images. The plugin uses [SRWieZ/thumbhash](https://github.com/SRWieZ/thumbhash) under the hood.

I used `php-scoper` to create a **scoped release asset**. This means I have a version of the plugin that is less likely to cause conflicts with other plugins. I plan to use this for the WP.org plugin directory.

## Installing via Composer

Another way to install WordPress plugins is through **Composer** and **Packagist**. A popular project for this is [Roots/Bedrock](https://roots.io/bedrock/).

- If someone uses the command `composer require hirasso/thumbhash-placeholders`, they will **not** get the scoped version. 
- They will get the **unscoped version** from the latest tag on the `main` branch.

If users get the unscoped version, it could lead to conflicts with other plugins or packages.

## My Main Question

What is the best way to scope a package when releasing it to Packagist?

- I can’t just use the `main` branch.
- Even with a scoped vendor directory, the `composer.json` file might still cause conflicts with other packages.

I hope this makes sense! I think many others have similar questions.

**Comment by theofidry** [comment]
>What is the best way to scope a package when releasing it to Packagist?

I think it depends on how you want to distribute your scoped asset: if it is as a PHAR only it can be part of the main project (like it is done with PHP-Scoper or Box). If you want to distribute a scoped Composer package, then you need a separate repository with the scoped code, like PHPStan does for instance. In that case you'll need a bit more work to configure the CI to push code to that scoped mirror-ish repository

**Reply by hirasso** [reply]
I want to release it via composer. Ok, seems like I need to learn even more GitHub tooling for creating that scoped version 🥴 ...or I wait until someone complains 😅

**Reply by hirasso** [reply]
An official recipe for pushing or pulling the `build` folder to a separate repo for composer would be amazing. That way, it would probably feel less hacky.

**Reply by theofidry** [reply]
A more automated way would be amazing, but unfortunately scoping in PHP is, unlike in JS for instance, a lot more finicky.

I couldn't really find a canonical way to do this yet as several projects that used PHP-Scoper tended to have a few steps done besides just shipping a "build" directory.

>I need to learn even more GitHub tooling for creating that scoped version

And I would need to learn how to do a GitHub recipe too 😅

**Reply by hirasso** [reply]
If you'd like me to present my current state to you and discuss possible solutions together, I'd love to hear from you via email (in my profile) so that we can jump on a call sometime soon. 

I can imagine talking it through could help a lot in this case 🙂

**Reply by hirasso** [reply]
I have created an [issue over at the Packagist rep](https://github.com/composer/packagist/issues/1483) to discuss this further. Seems like maintaining a separate repository currently is indeed the "official" way for releasing a scoped / built package.

**Comment by hirasso** [comment, marked as answer]
Wanted to get back here to post my current solution. People interested can clone my plugin from here: [hirasso/wp-thumbhash](https://github.com/hirasso/wp-thumbhash). 

## Decisions

- I decided to _not_ publish my plugin on the official WP.org plugin repository. Instead, I support plugin updates via composer or via the normal WordPress plugin update process / WP-CLI, powered by [Plugin Update Checker](https://github.com/YahnisElsts/plugin-update-checker).
- Even for installs via composer, my plugin is being scoped (for now). The packagist package is named `hirasso/wp-thumbhash`, but it actually points at [hirasso/wp-thumbhash-dist](https://github.com/hirasso/wp-thumbhash-dist).

## What I Did

I've created a small `cli.js` that takes care of all the things I need. I execute the relevant commands in my [ci and release workflows](https://github.com/hirasso/wp-thumbhash/tree/main/.github/workflows). They can also be executed locally through npm scripts. Here are the most important commands:

## Create a Scoped Release

```
config/cli/cli.js release:create
```

Scope all non-dev dependencies, inject other files and folders that do not have an `export-ignore` entry in the `.gitattributes` file. The scoped folder is created under `./scoped/my-plugin-slug/`. The command also creates a zip file `my-plugin-slug.zip` at the root level, to provide an installable release asset for [manual installs](https://github.com/hirasso/wp-thumbhash/blob/main/docs/INSTALLATION.md#manually).

## Test a Scoped Release 

```shell
config/cli/cli.js test:release
```

Run E2E tests against the scoped release folder using a @wordpress/env container and [Playwright](https://playwright.dev/):
## Push a Scoped Release to the Dist Repo

```shell
# clone the dist repo and checkout an empty root branch
config/cli/cli.js dist:prepare
# copy all scoped files into the dist repo folder, commit, tag and push
config/cli/cli.js dist:push
```

These commands assume your dist repo is named exactly like your source repo, with the suffix `-dist`. In my case, it's `hirasso/wp-thumbhash-dist`. Here is an [example tag](https://github.com/hirasso/wp-thumbhash-dist/releases/tag/v0.1.1), where GitHub's default zip file contains the scoped production-ready plugin files. This will be installed when running `composer require hirasso/wp-thumbhash`.

## Other Commands

There are a few other commands available that help me with other WordPress-specific things, like, for example, keeping the version in the main plugin file up to date (I'm handling versions/git tags and CHANGELOG entries using [@changesets/cli](https://github.com/changesets/changesets) and [@changesets/action](https://github.com/changesets/action)). 

Print all available commands to the console using the `help` command:

```shell
config/cli/cli.js help
```

```shell
Usage: cli.js <command>

Available commands:
  release:create – Create a scoped release
  version:patch – Patch the version in the main plugin file
  dist:prepare – Prepare the folder for pushing to the dist repo
  dist:push – Push the prepared dist folder to the dist repo
  test:dev – Run PHP and E2E tests against the unscoped development version of the plugin
  test:release – Run E2E tests against the scoped release version of the plugin
  help – Show available commands for this cli
```

## Notes

- I'm maintaining a separate [`composer.dist.json`](https://github.com/hirasso/wp-thumbhash/blob/main/composer.dist.json) that is being shipped with the scoped release instead of the actual [`composer.json`](https://github.com/hirasso/wp-thumbhash/blob/main/composer.json). It includes only the bare minimum required to be installed as a WordPress plugin. It **does not include any dependencies** except for the PHP version and `composer/installers`. Otherwise, installs via composer could still be blocked by version conflicts, even though all dependencies are scoped and shipped with my plugin.
- For excluding WP-specific globals I'm using [snicco/php-scoper-wordpress-excludes](https://github.com/snicco/php-scoper-wordpress-excludes). You can see how I'm doing it in my scoper.config.php [here](https://github.com/hirasso/wp-thumbhash/blob/31b65fa1e175088834ac6a3db050dc8b8e489772/config/scoper.config.php#L23) and [here](https://github.com/hirasso/wp-thumbhash/blob/31b65fa1e175088834ac6a3db050dc8b8e489772/config/scoper.config.php#L151-L165).
- I somehow ended up building my `cli` using JavaScript – it's a little easier for me to write. But of course, PHP would also be a viable option. Maybe even a better one, as scoping is so php-centric. Maybe using a tool like [Deployer](https://deployer.org/) would even be a good idea... because that is basically what we are doing: build, pack, deploy.
- Currently, no matter [how the plugin is installed](https://github.com/hirasso/wp-thumbhash/blob/main/docs/INSTALLATION.md) (composer, wp-cli, manual), **it will always install the scoped version**. That's because of a [version conflict I have with `snicco/better-wp-cli` and `psr/log`](https://github.com/snicco/snicco/issues/204).
- The scoped release will also be attached to the GitHub release as `wp-thumbhash.zip` from where it can be installed manually. It can update itself the classic WordPress way through an included instance of [Plugin Update Checker](https://github.com/YahnisElsts/plugin-update-checker). I would love if there was a way to point packagist towards a custom release asset zip file. If there was, I could skip the [dist repo](https://github.com/hirasso/wp-thumbhash-dist) altogether.

In the future, I might go the extra mile at some point and rewrite my plugin to not make use of `snicco/better-wp-cli`, so that I can get rid of the version conflicts with `psr/log` and support clean composer installs from the source. That will also force me to use [composer-dist-plugin](https://github.com/pixelfear/composer-dist-plugin) for my compiled JS and CSS assets, should I need to introduce a build step at some point 🤦🤦 ...but at least I would then be "doing it right", as discussed in depth in [this issue on the packagist repo](https://github.com/composer/packagist/issues/1483).

I'll mark this as "The Answer" for now, until someone else comes along and shows us how it's actually done :)

**Comment by theofidry** [comment]
Feel free to add a link to this entry in the docs, e.g. in the Wordpress support

**Comment by tombroucke** [comment]
I have been working on a working example for WordPress plugins. I came up with this: https://github.com/tombroucke/wp-plugin-php-scoper-example. The readme file contains all needed information. Feel free to suggest improvements ...

**Comment by owaisahmed5300** [comment]
Thanks for documenting your setup. I wanted to add one note for anyone following a similar approach.

If your plugin only needs exclusions for WordPress Core, `sniccowp/php-scoper-wordpress-excludes` is a good option. If it also depends on packages like WooCommerce, Action Scheduler, ACF, or WP-CLI, you'll likely need exclusions for those symbols as well.

I maintain **WPTechnix/php-scoper-wordpress-excludes**, which takes the same automated approach but extends it to supported WordPress ecosystem packages by generating exclusion lists from their latest upstream stubs.

https://github.com/WPTechnix/php-scoper-wordpress-excludes
